### PR TITLE
Clean up Pages & ViewModels

### PR DESCRIPTION
**Pull request by brianlagunas**
Dispose Pages and ViewModels when popping them off of the NavigationPage.NavigationStack or when popping them off of the Modal stack by calling IDisposable.  Also handles all children if applicable.

### CODE REVIEW

**Comment by brianlagunas** [comment]
This is to address #793

File: Source/Xamarin/Prism.Forms/Common/PageUtilities.cs [hunk 1]
```diff
@@ -21,5 +21,61 @@ public static class PageUtilities
...
+
+        public static void DisposePage(Page page)
+        {
+            try
+            {
+                DisposeChildren(page);
+
+                var viewModel = page.BindingContext as IDisposable;
+                if (viewModel != null)
+                    viewModel.Dispose();
```
Comment (joacar) [inline Source/Xamarin/Prism.Forms/Common/PageUtilities.cs]: Use null conditional-syntax instead -> viewModel?.Dipose()

**Comment by joacar** [comment]
Calling IDisposable twice should be on the applications responsibility to handle correctly - so this should pose no treat. 

If, as you say, Forms calls it and does so on a different thread then we are in for some trouble.

**Comment by brianlagunas** [comment]
When I get time, I might have to dig through Xamarin.Forms source code and see if IDisposable is called anywhere.

**Comment by brianlagunas** [comment]
So thinking about it, to properly implement IDisposable requires a big of repetitive code to do it properly.  This would require more knowledge on the developers part to guard themselves from issues related to an improper IDisposable implementation as well as add a little bloat to a ViewModel. Granted they could probably just put the majority in a base class.  The purpose of this feature is to allow devs to clean up Views/Viewmodels from events and other items.  So, I'm not 100% sure we should be using IDisposable here.  I'm not apposed to it, but I am not sold on it either.  I want to make it easy and worry free.  Thoughts?

**Comment by brianlagunas** [comment]
Any ideas on what the name of this interface should be?

**Comment by joacar** [comment]
Tricky... 

First thing that pops up in my head is something with lifecycle. But `IView{Model}LifecycleAware` doesn't  sound appealing. Then we have `IActiveAware` which, in a sense, handles lifecycle events.

**Comment by brianlagunas** [comment]
We can't use IActive aware, because that is about activation, not cleaning up because your view/vm has ben removed form the navigation stack.

Something like `IDestroy`, `IClean`, or `IPopped`

**Comment by brianlagunas** [comment]
I'm leaning towards `IDestroy`.  Also, keep in mind that this will be called on both the View and ViewModel :)

**Comment by joacar** [comment]
Didn't think to use `IActiveAware`, just wrote it to for "brain storming" purpose :)

`IPopped` and `IDestroy`  makes it to the final. 

`IDestroy` make me think about `OnDestroy` in Android so from cross-platform point of view maybe `IPopped` is more platform-independant name. But since it will be called on pages that aren't popped, `IDestroy` seem more suitable.

**Comment by brianlagunas** [comment]
I have also included a solution to #744 in this PR since they are so closely related.

**Pull request merged**